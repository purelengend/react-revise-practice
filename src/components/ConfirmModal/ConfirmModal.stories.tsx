import type { Meta, StoryObj } from "@storybook/react";
import { Center } from "@chakra-ui/react";
import { useArgs } from "@storybook/preview-api";

// Components
import ConfirmModal from ".";

const meta: Meta<typeof ConfirmModal> = {
  title: "Components/ConfirmModal",
  component: ConfirmModal,
  decorators: [
    (Story) => (
      <Center style={{ width: "100vh" }}>
        <Story />
      </Center>
    ),
  ],
  argTypes: {
    id: {
      description: "Id of the student to be confirmed",
    },
    isMutating: {
      description: "Determine whether the form is mutating or not",
    },
    isOpen: {
      description: "Determine whether the modal is open or not",
    },
    onClose: {
      description: "Function to close the modal",
    },
    onSubmit: {
      description: "Function to submit the form",
    },
  },

  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof ConfirmModal>;

export const Primary: Story = {
  args: {
    id: "0",
    isMutating: false,
    isOpen: true,
    onSubmit: () => {},
  },
  render: function Render(args) {
    const [, updateArgs] = useArgs<typeof args>();

    const onClose = () => {
      updateArgs({ isOpen: false });
    };

    return <ConfirmModal {...args} onClose={onClose} />;
  },
};
