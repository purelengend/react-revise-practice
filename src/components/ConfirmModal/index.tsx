import {
  Button,
  Input,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
  Text,
} from "@chakra-ui/react";
import { Controller, useForm } from "react-hook-form";

export type ConfirmFormProps = {
  id: string;
};
export type ConfirmModalProps = {
  id: string;
  title: string;
  isOpen: boolean;
  isMutating: boolean;
  onClose: () => void;
  onSubmit: (data: ConfirmFormProps) => void;
};

const ConfirmModal = ({
  id,
  title,
  isOpen,
  isMutating,
  onClose,
  onSubmit,
}: ConfirmModalProps) => {
  const { control, handleSubmit } = useForm<ConfirmFormProps>({
    defaultValues: {
      id,
    },
  });

  return (
    <Modal
      autoFocus={false}
      returnFocusOnClose={false}
      isOpen={isOpen}
      onClose={onClose}
    >
      <ModalOverlay />
      <ModalContent
        bg="white"
        px={0}
        minW={{
          base: "auto",
          md: 169,
        }}
      >
        <ModalHeader
          textAlign="center"
          py={4}
          borderBottom="1px solid"
          borderBottomColor="white.300"
          fontSize="lg"
        >
          {title}
        </ModalHeader>
        <form id="#confirm-form" noValidate onSubmit={handleSubmit(onSubmit)}>
          <ModalBody py={4}>
            <Text color="gray.200" fontSize="md" textAlign="center">
              This action cannot be undone, so please be careful with this
              action.
            </Text>
            <Controller
              name="id"
              control={control}
              render={({ field }) => <Input {...field} hidden isDisabled />}
            />
          </ModalBody>

          <ModalFooter display="flex" gap={5}>
            <Button
              type="button"
              w={30}
              px={7.5}
              py={4}
              onClick={onClose}
              bg="blackAlpha.100"
              _hover={{
                bg: "blackAlpha.300",
              }}
              isDisabled={isMutating}
            >
              Cancel
            </Button>
            <Button
              type="submit"
              w={30}
              px={7.5}
              py={4}
              bg="yellow.200"
              _hover={{
                bg: "orange.400",
              }}
              isDisabled={isMutating}
            >
              Yes
            </Button>
          </ModalFooter>
        </form>
      </ModalContent>
    </Modal>
  );
};

export default ConfirmModal;
