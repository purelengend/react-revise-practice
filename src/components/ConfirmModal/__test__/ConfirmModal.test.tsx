import { screen, renderHook, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useDisclosure } from "@chakra-ui/react";

// Component
import ConfirmModal from "..";

describe("ConfirmModal test cases", () => {
  const mockOnSubmit = jest.fn();

  window.URL.createObjectURL = jest.fn();

  const {
    result: {
      current: { isOpen, onClose },
    },
  } = renderHook(() =>
    useDisclosure({
      defaultIsOpen: true,
    }),
  );

  const mockId: string = "0";

  const setup = () => {
    userEvent.setup();
    return render(
      <ConfirmModal
        id={mockId}
        title=""
        isOpen={isOpen}
        isMutating={false}
        onClose={onClose}
        onSubmit={mockOnSubmit}
      />,
    );
  };
  it("should render correctly", () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it("should invoke submit function when clicking submit button", async () => {
    setup();

    const submitBtn = screen.getByRole("button", {
      name: /yes/i,
    });

    await userEvent.click(submitBtn);

    expect(mockOnSubmit).toHaveBeenCalled();
  });
});
