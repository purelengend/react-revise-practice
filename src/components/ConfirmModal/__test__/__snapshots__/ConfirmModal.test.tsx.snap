// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal test cases should render correctly 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
